clusterloader2/pkg/measurement/common/pod_command.go

```diff
@@ -0,0 +1,697 @@
+//     Pods are selected using `params.labelSelector`, must contain `params.container`, and must be in a running
+//     state. If no applicable pods are available, then no step is performed for the tick.
+//  2. For each pod, spin a goroutine which will run all configured commands in the pod.
+//  3. For each command, spin a goroutine to handle running the command.
+//  4. If the command returns non-zero, this will be reflected in the associated measurement.
+//  5. If an error occurs while attempting to execute the command, such as a temporary network failure,
+//     and no such error has occurred yet during the course of the measurement, then the error will be saved and
+//     returned during the gather step if `params.failOnError` is true. For example, if multiple commands fail due
+//     to a network failure, then only this first error will be saved and returned in the gather step when
+//     `params.failOnError` is set. The rest of the observed errors will be logged.
+//
+// The following measurements are produced during the gather step:
+//
+//  1. One summary measurement, which includes information on all executed commands, such as if the command
+//     took longer than `params.timeout`, the command's RC, and the pod the command was executed on.
+//  2. One measurement for each command's non-empty stdout and stderr.
+//  3. One measurement containing statistics, such as the number of commands executed, the number of errors observed,
+//     and the number of non-zero RCs.
+type podPeriodicCommandMeasurement struct {
+	clientset  kubernetes.Interface
+	restConfig *rest.Config
+	params     *podPeriodicCommandMeasurementParams
+	isRunning  bool
+	// stopCh is closed when stop() is called.
+	stopCh chan struct{}
+	// doneCh is closed after stopCh is closed and all in progress commands have finished.
+	doneCh  chan struct{}
+	results []*runAllCommandsResult
+	// resultsLock needs to be held to modify and read the results field.
+	resultsLock *sync.Mutex
+	informer    cache.SharedInformer
+	// err is returned during the gather step.
+	err error
+	// errOnce needs to be used to modify the err field.
+	errOnce sync.Once
+	stats   *stats
+	// statsLock needs to be hepl to modify and read the stats field.
+	statsLock *sync.Mutex
+}
+
+// isApplicablePod checks if a pod is a viable candidate for running a command on.
+func (p *podPeriodicCommandMeasurement) isApplicablePod(pod *v1.Pod) bool {
+	hasContainer := false
+	for _, c := range pod.Spec.Containers {
+		if c.Name == p.params.Container {
+			hasContainer = true
+
+			break
+		}
+	}
+
+	if !hasContainer {
+		return false
+	}
+
+	if pod.Status.Phase != v1.PodRunning {
+		return false
+	}
+
+	isReady := false
+	for _, c := range pod.Status.Conditions {
+		if c.Type == v1.PodReady && c.Status == v1.ConditionTrue {
+			isReady = true
+
+			break
+		}
+	}
+
+	if !isReady {
+		return false
+	}
+
+	for _, c := range pod.Status.ContainerStatuses {
```

Comment:
We can probably get rid of checking container status as `PodReady` condition checks for all containers to be in a ready state: https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-readiness-status

Reply:
Oh great! TIL. Removed.